types/sap__approuter/index.d.ts

```diff
@@ -0,0 +1,389 @@
+         *               done(null, incomingResponse);
+         *           } else {
+         *               done('An error occurred in backend, returned status ' + ctx.outgoingResponse.statusCode, ctx.incomingResponse);
+         *           }
+         *       };
+         *       next();
+         *   });
+         *   ar.start();
+         * @param ctx - Context object for the request handler
+         * @param {AppRouterIncomingMessage~afterRequestHandlerCallback} done - A callback function that receives (optionally) and error and the modified incomingResponse
+         */
+        afterRequestHandler?: (
+            ctx: AfterRequestHandlerContext,
+            done: (error: Error | string | undefined | null, incomingResponse: ServerResponse) => void,
+        ) => void;
+
+        /**
+         * A function that can be added to the request object - for example in a "first" or "beforeRequestHandler" extension.
+         * If exists, this function will be called by the standard application router when a backend connection timeout occurs.
+         *
+         * @param req - the request object
+         * @param done - a callback function that doesn't return any parameter
+         */
+        backendTimeout?: (req: IncomingMessage, done: () => void) => void;
+
+        /**
+         * Callback for {@link afterRequestHandler}
+         *
+         * @param error - Error object in case of error
+         * @param incomingResponse - Modified incomingResponse
+         */
+    }
+
+    /**
+     * A handler for requests, called Middleware.
+     *
+     * @param request - Data of the incoming Request
+     * @param response - Object for the outgoing Response; can be manipulated to alter the response
+     * @param next - Call next to give control back to Application Router Middleware
+     */
+    type MiddlewareHandler = (request: AppRouterIncomingMessage, response: ServerResponse, next: () => void) => void;
+
+    interface Extensions {
+        /**
+         * Describes the middleware provided by this extension
+         * If a path is given, the handler is only invoked on these paths; otherwise it's invoked on all requests.
+         */
+        insertMiddleware: {
+            /** A MiddlewareSlot before the first application router middleware */
+            first?: Array<MiddlewareHandler | { path: string; handler: MiddlewareHandler }>;
+            /** A MiddlewareSlot before the standard application router request handling */
+            beforeRequestHandler?: Array<MiddlewareHandler | { path: string; handler: MiddlewareHandler }>;
+            /** A MiddlewareSlot before the standard application router error handling */
+            beforeErrorHandler?: Array<MiddlewareHandler | { path: string; handler: MiddlewareHandler }>;
+        };
+    }
+
+    type RouterConfig = unknown;
+
+    interface StartOptions {
+        /** A TCP port the application router will listen to */
+        port?: string;
```

Comment:
One last suggestion; you probably want to add `| undefined` to all of these optional properties to support `exactOptionalPropertyTypes`, unless they really cannot have an explicit `undefined`.

See: [Playground Link](https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.2#code/PTAEHUEsBcAtQKYA8CGBjaB5ADtSB7AOxQBsAFAJ32wQugE8AVemgZ0WICMSEATAGlDMaAZTQVIuALAAoEKFgp2raBIy1QFAK492KKlsK9QAd0XRQ9fFtBoUhUKwQX7oaniKlZ87FRp16UGh8ADpZbzAAMXwKRFQAW2weQThIdkhCaFoAM3QEUF4ENBJ9BHY4FAs4Wny00BQI0F9qWgZTWEg0eDsHTnyifuygk3xHVQyAc1YALlAAcl59AGs5t1i5kkgJ2Gg5xpjQGEdYaxJjQnwLPsOHardOACsi6DCZWQysily0fIBVJwoABEELkdNB2ABvWSgUDyRiwfIoThOQg-NxDFCgABupC0+QoCF8ZQQmXYc1Y9BUCHiexkMLQ+BIMXh1IQmCxtAkhQA-LMAESLChLPmgAA+oD5m220D5AG5ZABfcJyMBQODWKoddjZEoTDhIngCIIIgn1U1wAn5HEkPHsMydeBWGw9RpOCwAAwZTIoLPibI5FC5CHdQVGfX5guFgklWx2IvsxndhkK2QyfHdr0aImceEIeru1rxodASaMILTvBDJkgJBI9Vr+BMoHi+BUoAAUigcWIJLhNIY8H7Gl0ikttQc7sg0llUflgqBcjXBPauod2CDss9IBySIFciQKa95AAJRsIAMpE0Icn1BdaABe97TrHKo1S7D6iixBAoISECIKctCBgAhCEabIDk9RlmQRP12U5SBChDAB9Xh8DKQg5mgZCbmKLRChLZMgPTTMVX-fIdRQPV3WQdAsFwUDSEoFoAmEMoQ3iFAljKIQWAQHtJBcNA0C0ChKgQXdGk46BVzuFMMhAogmiob9CmMJR6gcdxGLrZp-AYUjCmKUo71RDwHAmZx-lobNoFzKYAAoAEpZmsoEQRQMFWHlN4ZAZQg2zdez2AAXlASzoDc2zgucnygsmVgQi9GDWXgwNEPyMKBWWOVZHivNEuSn1YP9BCCKyqU4x8xoACEtAsC5oGmPKcwSpLoOK1KAyDUAwqI1NCD4aqyMoH9izuSjqNojAcHMpi-FaJg+NYd1LzqEwNMappW1YSBuAQEIgA)

Reply:
I tested it but set the option to undefined throws an error when using the library:
In this case `VError: options/port: Invalid type: undefined (expected string/integer)`. In this case I think it would be misleading

Reply:
@jakebailey If I understand it correctly it makes only sense if the function accepts undefined properties and handles them same as not set properties